src/core/request.js

```diff
@@ -161,92 +179,102 @@ export default function request(options) {
-                                    return resolve(response);
-                                }
+                        if (xhr.status === 200 || (response && response.success === true)) {
+                            // there's some data
+                            return resolve(response);
+                        }
 
-                                //the server has handled the error
-                                reject(
-                                    createError(
-                                        response,
-                                        __('The server has sent an empty response'),
-                                        xhr.status,
-                                        xhr.readyState > 0
-                                    )
-                                );
-                            })
-                            .catch(function(error) {
-                                logger.error(error);
-                                reject(createError(response, error, xhr.status, xhr.readyState > 0));
-                            });
+                        //the server has handled the error
+                        reject(
+                            createError(
+                                response,
+                                __('The server has sent an empty response'),
+                                xhr.status,
+                                xhr.readyState > 0
+                            )
+                        );
                     })
-                    .fail(function(xhr, textStatus, errorThrown) {
-                        var response;
-                        try {
-                            response = JSON.parse(xhr.responseText);
-                        } catch (parseErr) {
-                            response = {};
-                        }
+                    .catch(error => {
+                        logger.error(error);
+                        reject(createError(response, error, xhr.status, xhr.readyState > 0));
+                    });
+                })
+                .fail((xhr, textStatus, errorThrown) => {
+                    let response;
+                    try {
+                        response = JSON.parse(xhr.responseText);
+                    } catch (parseErr) {
+                        response = {};
+                    }
+
+                    const responseExtras = {
+                        success: false,
+                        source: 'network',
+                        cause: options.url,
+                        purpose: 'proxy',
+                        context: this,
+                        code: xhr.status,
+                        sent: xhr.readyState > 0,
+                        type: 'error',
+                        textStatus: textStatus,
+                        message: errorThrown || __('An error occurred!')
+                    }
+
+                    const enhancedResponse = {...response, ...responseExtras};
```

Comment:
Not sure about this one. It will act as `Object.assign` (or `_.assign()`), but not as `_.defaults()`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#Spread_in_object_literals

Reply:
Good catch. I had a small doubt about `_.defaults` vs object spread, but getting the arguments in the right order makes a big difference!